File: CS/FilteredData/MyGridControl/MyTreeListView.cs
using DevExpress.Xpf.Grid;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Collections.Specialized;
using System.Linq;
using System.Windows;

namespace FilteredData.MyGridControl {
    class MyTreeListView : TreeListView {
        bool ChangedFromViewModel;
        public ObservableCollection<object> ForceUnfiltered;
        public object NodeToDelete;

        public static readonly DependencyProperty VisibleDataProperty = DependencyProperty.Register("VisibleData", typeof(ObservableCollection<object>), typeof(MyTreeListView), new UIPropertyMetadata(FilteredDataPropertyChanged));
        public ObservableCollection<object> VisibleData {
            get { return (ObservableCollection<object>)GetValue(VisibleDataProperty); }
            set { SetValue(VisibleDataProperty, value); }
        }
        private static void FilteredDataPropertyChanged(DependencyObject source, DependencyPropertyChangedEventArgs e) {
            MyTreeListView view = source as MyTreeListView;
            view.ChangedFromViewModel = true;
            view.ForceUnfiltered = new ObservableCollection<object>();
            ObservableCollection<object> list = e.NewValue as ObservableCollection<object>;
            view.DataControl.RefreshData();
            list.CollectionChanged += new System.Collections.Specialized.NotifyCollectionChangedEventHandler(view.list_CollectionChanged);
            view.CustomNodeFilter += new DevExpress.Xpf.Grid.TreeList.TreeListNodeFilterEventHandler(view.view_CustomNodeFilter);
        }

        public void view_CustomNodeFilter(object sender, DevExpress.Xpf.Grid.TreeList.TreeListNodeFilterEventArgs e) {
            if (ForceUnfiltered.Contains(e.Node.Content))
                e.Visible = true;
            e.Handled = true;
        }

        public void list_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) {
            if (!ChangedFromViewModel)
                return;
            if (e.Action == NotifyCollectionChangedAction.Add && !ForceUnfiltered.Contains(e.NewItems[0])) {
                ForceUnfiltered.Add(e.NewItems[0]);
                this.DataControl.RefreshData();
            }
            if (e.Action == NotifyCollectionChangedAction.Remove && ForceUnfiltered.Contains(e.OldItems[0])) {
                ForceUnfiltered.Remove(e.OldItems[0]);
                NodeToDelete = e.OldItems[0];
                this.DataControl.RefreshData();
            }
        }

        void dataProvider_RefreshFilteredData(object sender, MyFilterEventArgs e) {
            ObservableCollection<object> col = VisibleData;
            if (col == null) {
                return;
            }
            bool f = false;
            if (f)
                col.Clear();
            ChangedFromViewModel = false;
            if (e.IsVisible)
                if (col.Contains(e.Node.Content)) {
                    ChangedFromViewModel = true;
                    return;
                }
                else
                    col.Add(e.Node.Content);
            else {
                if (col.Contains(e.Node.Content) || NodeToDelete == e.Node.Content) {
                    List<object> objectsToRemove = new List<object>();
                    objectsToRemove.AddRange(new TreeListNodeIterator(e.Node.Nodes).Select<TreeListNode, object>(n => n.Content));
                    objectsToRemove.Add(e.Node.Content);
                    foreach (object o in objectsToRemove)
                        col.Remove(o);
                    NodeToDelete = null;
                }
            }
            ChangedFromViewModel = true;
        }

        protected override DevExpress.Xpf.Grid.TreeList.TreeListDataProvider CreateDataProvider() {
            MyTreeListDataProvider dataProvider = new MyTreeListDataProvider(this);
            dataProvider.FilteredData += new MyTreeListDataProvider.RefreshFilteredDataEventHandler(dataProvider_RefreshFilteredData);
            return dataProvider;
        }

    }
}

File: VB/FilteredData/MyGridControl/MyTreeListView.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.Xpf.Grid
Imports System.Collections.ObjectModel
Imports System.Windows
Imports System.Collections.Specialized

Namespace FilteredData.MyGridControl
	Friend Class MyTreeListView
		Inherits TreeListView
		Private ChangedFromViewModel As Boolean
		Public ForceUnfiltered As ObservableCollection(Of Object)
		Public NodeToDelete As Object

        Public Shared ReadOnly VisibleDataProperty As DependencyProperty = DependencyProperty.Register("VisibleData", GetType(ObservableCollection(Of Object)), GetType(MyTreeListView), New UIPropertyMetadata(AddressOf FilteredDataPropertyChanged))
		Public Property VisibleData() As ObservableCollection(Of Object)
			Get
				Return CType(GetValue(VisibleDataProperty), ObservableCollection(Of Object))
			End Get
			Set(ByVal value As ObservableCollection(Of Object))
				SetValue(VisibleDataProperty, value)
			End Set
		End Property
		Private Shared Sub FilteredDataPropertyChanged(ByVal source As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
			Dim view As MyTreeListView = TryCast(source, MyTreeListView)
			view.ChangedFromViewModel = True
			view.ForceUnfiltered = New ObservableCollection(Of Object)()
			Dim list As ObservableCollection(Of Object) = TryCast(e.NewValue, ObservableCollection(Of Object))
			view.DataControl.RefreshData()
			AddHandler list.CollectionChanged, AddressOf view.list_CollectionChanged
			AddHandler view.CustomNodeFilter, AddressOf view.view_CustomNodeFilter
		End Sub

		Public Sub view_CustomNodeFilter(ByVal sender As Object, ByVal e As DevExpress.Xpf.Grid.TreeList.TreeListNodeFilterEventArgs)
			If ForceUnfiltered.Contains(e.Node.Content) Then
				e.Visible = True
			End If
			e.Handled = True
		End Sub

		Public Sub list_CollectionChanged(ByVal sender As Object, ByVal e As System.Collections.Specialized.NotifyCollectionChangedEventArgs)
			If (Not ChangedFromViewModel) Then
				Return
			End If
			If e.Action = NotifyCollectionChangedAction.Add AndAlso (Not ForceUnfiltered.Contains(e.NewItems(0))) Then
				ForceUnfiltered.Add(e.NewItems(0))
				Me.DataControl.RefreshData()
			End If
			If e.Action = NotifyCollectionChangedAction.Remove AndAlso ForceUnfiltered.Contains(e.OldItems(0)) Then
				ForceUnfiltered.Remove(e.OldItems(0))
				NodeToDelete = e.OldItems(0)
				Me.DataControl.RefreshData()
			End If
		End Sub

		Private Sub dataProvider_RefreshFilteredData(ByVal sender As Object, ByVal e As MyFilterEventArgs)
			Dim col As ObservableCollection(Of Object) = VisibleData
			If col Is Nothing Then
				Return
			End If
			Dim f As Boolean = False
			If f Then
				col.Clear()
			End If
			ChangedFromViewModel = False
			If e.IsVisible Then
				If col.Contains(e.Node.Content) Then
					ChangedFromViewModel = True
					Return
				Else
					col.Add(e.Node.Content)
				End If
			Else
				If col.Contains(e.Node.Content) OrElse NodeToDelete Is e.Node.Content Then
					Dim objectsToRemove As New List(Of Object)()
					Dim nodeIterator As New TreeListNodeIterator(e.Node.Nodes, False)
					Do While nodeIterator.MoveNext()
						objectsToRemove.Add(nodeIterator.Current.Content)
					Loop
					objectsToRemove.Add(e.Node.Content)
					For Each o As Object In objectsToRemove
						col.Remove(o)
					Next o
					NodeToDelete = Nothing
				End If
			End If
			ChangedFromViewModel = True
		End Sub

		Protected Overrides Function CreateDataProvider() As DevExpress.Xpf.Grid.TreeList.TreeListDataProvider
			Dim dataProvider As New MyTreeListDataProvider(Me)
			AddHandler dataProvider.FilteredData, AddressOf dataProvider_RefreshFilteredData
			Return dataProvider
		End Function

	End Class
End Namespace
